cpp/src/strings/char_types/char_types.cu

```diff
@@ -83,6 +83,107 @@ std::unique_ptr<column> all_characters_of_type(
   return results;
 }
 
+namespace {
+
+/**
+ * @brief Removes individual characters from a strings column based on character type.
+ *
+ * Types to remove are specified by `types_to_filter` OR
+ * types to not remove are specified by `types_to_keep`.
+ *
+ * This is called twice. The first pass calculates the size of each output string.
+ * The final pass copies the results to the output strings column memory.
+ */
+struct filter_chars_fn {
+  column_device_view const d_column;
+  character_flags_table_type const* d_flags;
+  string_character_types const types_to_filter;
+  string_character_types const types_to_keep;
+  string_view const d_replacement;  ///< optional replacement for removed characters
+  int32_t* d_offsets{};             ///< size of the output string stored here during first pass
+  char* d_chars{};                  ///< this is null only during the first pass
+
+  /**
+   * @brief Returns true if the given character should be replaced.
+   */
+  __device__ bool replace_char(char_utf8 ch)
+  {
+    auto const code_point = detail::utf8_to_codepoint(ch);
+    auto const flag       = code_point <= 0x00FFFF ? d_flags[code_point] : 0;
+    if (flag == 0)  // all types pass unless specifically identified
+      return (types_to_filter == ALL_TYPES);
+    if (types_to_keep == ALL_TYPES)  // filter case
+      return (types_to_filter & flag) != 0;
+    return (types_to_keep & flag) == 0;  // keep case
+  }
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_column.is_null(idx)) {
+      if (!d_chars) d_offsets[idx] = 0;
+      return;
+    }
```

Comment:
I think it's possible to get more performance out of functions such as these if we operate per individual characters using string offsets as a guide, rather than operating per-string - unless all strings are all valid and all the same length. Per-string is the approach we use elsewhere, though. :man_shrugging: